### PR TITLE
[llvm] [ir] Wrap struct-for's single coordinate computation into another function

### PR DESCRIPTION
**Pull request by k-ye**
I'm taking the first approach here, but I wonder if this can be inlined automatically? @yuanming-hu 

Related issue = #1377

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Comment by archibate** [comment]
I suddenly realized that, if Metal use gird-stride loop, using continue as return could be fatal?

**Comment by k-ye** [comment]
> I suddenly realized that, if Metal use gird-stride loop, using continue as return could be fatal?

Thanks, but no.. Metal already wraps the computation for a single index inside a function, in both range-for and struct-for. This is the Metal source for the added example:

```cpp
void mtl_k0001_run_c4_0_0_func(
    device byte* root_addr,
    device byte* global_tmps_addr,
    device byte* runtime_addr,
    device byte* print_addr,
    thread const ListgenElement& listgen_elem_,
    const int linear_loop_idx_) {
  device Runtime *runtime_ = reinterpret_cast<device Runtime *>(runtime_addr);
  device RandState* rand_state_ = reinterpret_cast<device RandState*>(runtime_->rand_seeds + (linear_loop_idx_ % 65536));
  device auto* print_alloc_ = reinterpret_cast<device PrintMsgAllocator*>(print_addr);
  const int tmp7 = listgen_elem_.coords[0];
  constexpr int32_t tmp8 = 0;
  const int32_t tmp9 = -(tmp7 == tmp8);
  constexpr int32_t tmp10 = 1;
  const int32_t tmp11 = (tmp9 & tmp10);
  constexpr int32_t tmp12 = 17;
  const int32_t tmp13 = -(tmp7 == tmp12);
  const int32_t tmp15 = (tmp13 & tmp10);
  const int32_t tmp16 = (tmp11 | tmp15);
  if (tmp16) {
    return;  <-- derived from `continue`
  } else {
  }
  S0 tmp88(root_addr);
  S0_ch tmp90 = tmp88.children(tmp8);
  S3 tmp91 = tmp90.get1();
  S3_ch tmp94 = tmp91.children(tmp7);
  device int32_t* tmp95 = tmp94.get0().val;
  *tmp95 = tmp10;
}


kernel void mtl_k0001_run_c4_0_0(
    device byte* root_addr [[buffer(0)]],
    device byte* global_tmps_addr [[buffer(1)]],
    device byte* runtime_addr [[buffer(2)]],
    device byte* print_addr [[buffer(3)]],
    const uint ugrid_size_ [[threads_per_grid]],
    const uint utid_ [[thread_position_in_grid]]) {
  // struct_for
  device Runtime *runtime_ = reinterpret_cast<device Runtime *>(runtime_addr);
  device MemoryAllocator *mem_alloc_ = reinterpret_cast<device MemoryAllocator *>(runtime_ + 1);
  ListManager parent_list;
  parent_list.lm_data = (runtime_->snode_lists + 1);
  parent_list.mem_alloc = mem_alloc_;
  const SNodeMeta parent_meta = runtime_->snode_metas[1];
  const int child_stride = parent_meta.element_stride;
  const int child_num_slots = parent_meta.num_slots;
  for (int ii = utid_; ii < 32; ii += ugrid_size_) {
    const int parent_idx_ = (ii / child_num_slots);
    if (parent_idx_ >= num_active(&parent_list)) return;
    const int child_idx_ = (ii % child_num_slots);
    const auto parent_elem_ = get<ListgenElement>(&parent_list, parent_idx_);
    ListgenElement listgen_elem_;
    listgen_elem_.root_mem_offset = parent_elem_.root_mem_offset + child_idx_ * child_stride;
    if (!is_active(root_addr + listgen_elem_.root_mem_offset, parent_meta, child_idx_)) continue;
    refine_coordinates(parent_elem_, runtime_->snode_extractors[1], child_idx_, &listgen_elem_);
    mtl_k0001_run_c4_0_0_func(root_addr, global_tmps_addr, runtime_addr, print_addr, listgen_elem_, ii);  <-- function for a single index
  }
}

```

File: taichi/codegen/codegen_llvm.cpp [hunk 1]
```diff
@@ -1368,7 +1368,21 @@ void CodeGenLLVM::create_offload_struct_for(OffloadedStmt *stmt, bool spmd) {
...
+      {
+        auto single_index_guard = get_function_creation_guard({
+            runtime_context_ty,  // The first arg is expected to be Context.
+            llvm::PointerType::get(physical_coordinate_ty, 0),
+        });
+        single_coord_func = single_index_guard.body;
+        current_coordinates = get_arg(1);
+        stmt->body->accept(this);
+      }
+      create_call(single_coord_func, {get_context(), new_coordinates});
```
Comment (archibate) [inline taichi/codegen/codegen_llvm.cpp:1385]: I'm afraid if this could harm performance?

File: tests/python/test_continue.py [hunk 2]
```diff
@@ -146,3 +146,28 @@ def run(a: ti.i32):
     assert x[0] == 1
     run(0)
     assert x[0] == 0
+
+
+@ti.all_archs
```
Comment (archibate) [inline tests/python/test_continue.py:151]:
```suggestion
@ti.archs_excluding(ti.opengl)
```

**Pull request closed without merging**